### DISCUSSION TITLE
AS1000_CONTROL_PAD_ Events

### DISCUSSION THREAD

**Discussion by bstudtma** [Pomax/msfs-simconnect-api-wrapper#29] [Q&A]
First of all, this is really impressive! Cool to see it in action. I tried to trigger the AS1000_CONTROL_PAD_* H events (new in SU14) from MSFS.trigger('AS1000_CONTROL_PAD_1",1) but noticed the API logged an error. Any idea if something like that would be possible? I was thinking about creating a web UI for a tablet to with buttons for the Cirrus Perspective+ using your API.

Thanks!

**Comment by Pomax** [comment]
What was the error?

**Comment by bstudtma** [comment]
On the API server:

Triggering sim event AS1000_CONTROL_PAD_1
RecvException { exception: 7, sendId: 595, index: 2 }
RecvException { exception: 1, sendId: 596, index: 2 }

Note that I triggered it from the browser running "are we flying". Thanks!

**Comment by Pomax** [comment]
exception 7 is `SIMCONNECT_EXCEPTION_NAME_UNRECOGNIZED`, meaning that the simconnect DLL doesn't know that variable. Are you sure it got added in SU14? (I don't see any mention of it when I search https://docs.flightsimulator.com/html/Introduction/Introduction.htm)

**Reply by bstudtma** [reply]
I was basing it off the RNs here: https://forums.flightsimulator.com/t/release-notes-sim-update-14-1-35-21-0-available-now/619902

https://devsupport.flightsimulator.com/t/su14-beta-g1000-nxi-control-pad-gcu-events-list/6754

https://hubhop.mobiflight.com/preset/?simType=msfs2020&id=1675c57e-bca5-43d4-880d-0a6b9a7e67a6

"Added support for hardware keyboard with new AS1000_CONTROL_PAD_ H events."

Perhaps that doesn't necessarily mean simconnect was updated. I'm not sure. Do I need to update the DLL in any of the project directories by chance? Are H events not supported?

**Reply by Pomax** [reply]
node-simconnect should be using MSFS's own simconnect, so I don't _think_ you need to do anything there. I don't know if H events are considered special in node-simconnect, but that might be something that @EvenAR probably knows.

**Reply by EvenAR** [reply]
Yes, it seems like H events works differently from the standard sim events. If I'm not mistaken these type of events were not accessible in SimConnect until [SU13](https://www.flightsimulator.com/release-notes-sim-update-13-1-34-16-0-available-now/). To trigger one of these events you will need to know the event's "hash id" which you can get by calling [`SimConnect_EnumerateInputEvents`](https://docs.flightsimulator.com/html/Programming_Tools/SimConnect/API_Reference/InputEvents/SimConnect_EnumerateInputEvents.htm) after the aircraft is loaded. These new APIs were [recently implemented in node-simconnect](https://github.com/EvenAR/node-simconnect/releases/tag/3.6.0).

Using node-simconnect I was able to trigger `AS1000_ControlPad_A` like this:

```ts
import { open, Protocol } from 'node-simconnect';

open('My app', Protocol.KittyHawk)
    .then(({ recvOpen, handle }) => {
        console.log('Connected: ', recvOpen);

        let allInputEvents: {
            name: string;
            idHash: Long;
            params?: string;
        }[] = [];

        handle.on('inputEventsList', ({inputEventDescriptors, entryNumber, outOf}) => {
            inputEventDescriptors.forEach(e => {
                allInputEvents.push({
                    name: e.name,
                    idHash: e.inputEventIdHash,
                });
            });
            if (entryNumber === outOf - 1) { // Did we receive the last page?
                const controlPadA = allInputEvents.find(inputEvent => inputEvent.name === "AS1000_ControlPad_A")
                if (controlPadA) {
                    handle.setInputEvent(controlPadA.idHash, 1);
                }
            }
        });

        handle.on('exception', ex => console.log('Exception', ex));

        handle.enumerateInputEvents(0 /* request-id */);
    })
    .catch(error => {
        console.log('Failed to connect', error);
    });
```

**Reply by EvenAR** [reply]
Rereading your question, @bstudtma, I noticed that the event name you mentioned, `AS1000_CONTROL_PAD_1`, is different from the one I got from `enumerateInputEvents` (`AS1000_ControlPad_1`). It seems like I might be wrong about H events and that `SimConnect_EnumerateInputEvents` is for something else, but the important thing is that you should still be able to control the G1000 with them. 😅

**Reply by bstudtma** [reply]
Thanks! I'm getting an error compiling the ts. 

"Cannot find name 'Long'."

Does this require a previous version of ts? Oddly enough, it was working for a bit but I tried to recompile and started seeing this error. I'm probably doing something silly - not too familiar with TS.

**Reply by bstudtma** [reply]
I managed to avoid TS by changing it as shown below and it seems to work again. I need to learn TS one of these days - I'm not sure why it won't compile. If you know offhand what the "Cannot find name 'Long'." error is about I'd certainly be interested.

Also, would it be possible to use these new functions with msfs-simconnect-api-wrapper? Or should I submit that as an issue/idea?

let sc = require('node-simconnect');
let open = sc.open;
let Protocol = sc.Protocol;
....

        let allInputEvents = [];

**Reply by Pomax** [reply]
I can try to add them, but then it's a good idea to file an issue for that. I'm currently writing the last words on the updated "are we flying" tutorial so I should have some time to focus on this library again pretty soon.

(it's double the length, and vastly better than before >_>)

**Reply by bstudtma** [reply]
I was able to implement those new node-simmconnect functions in a clone of msfs-simconnect-api-wrapper and it's working great for the most part. I can put in a pull request if that would be helpful. I was able to add most of the console buttons on the SR22T to a webpage (cloned 'are we flying' and created a sr22t.html and .js using the same APIs). I'm running the webpage on an iPad and can control the keypad and autopilot without moving the view. I'm struggling a with knobs still. Considering these events aren't documented, it's hard to know exactly what to send. AS1000_ControlPad_1_Heading sometimes makes it increment and sometimes decrements which seems strange. I can probably use a different sim connect event for changing the heading bug, but some of the other FMS knobs might not be so easy.

**Reply by Pomax** [reply]
PRs always welcome! Just remember to still file a real issue though, so it can be tied to that.

**Reply by Pomax** [reply]
as for the heading bug, since that's inherently tied to the plane, as a last resort you should be able to look up and then explicitly inc/dec the `AUTOPILOT_HEADING_LOCK_DIR` value, which changes the "locked in" heading value (after making sure `AUTOPILOT_HEADING_LOCK` is set to 1, otherwise no amount of changing that will do much =D)

**Reply by EvenAR** [reply]
> I was able to implement those new node-simmconnect functions in a clone of msfs-simconnect-api-wrapper and it's working great for the most part. I can put in a pull request if that would be helpful. I was able to add most of the console buttons on the SR22T to a webpage (cloned 'are we flying' and created a sr22t.html and .js using the same APIs). I'm running the webpage on an iPad and can control the keypad and autopilot without moving the view. I'm struggling a with knobs still. Considering these events aren't documented, it's hard to know exactly what to send. AS1000_ControlPad_1_Heading sometimes makes it increment and sometimes decrements which seems strange. I can probably use a different sim connect event for changing the heading bug, but some of the other FMS knobs might not be so easy.

Have you tried calling ```enumerateInputEventParams``` for the heading input event? Perhaps it requires some input parameters?

**Reply by EvenAR** [reply]
Nevermind, there was a bug in node-simconnect.. Please try the newly released v3.6.2 @bstudtma

**Reply by bstudtma** [reply]
Thanks @EvenAR. I noticed that last night and was going to submit an issue. I'll try your update tonight. I'm not sure if all numeric parameters are float64, but that seemed to be the case for the SR22T at least. I'll keep you posted. Thanks again.

**Reply by EvenAR** [reply]
I'm struggling to find any documentation or sample code which confirms the alllowed `SetInputEvent`-types and size values, but my assumption is that it is limited to the [`SIMCONNECT_INPUT_EVENT_TYPE`](https://docs.flightsimulator.com/flighting/html/Programming_Tools/SimConnect/API_Reference/Structures_And_Enumerations/SIMCONNECT_INPUT_EVENT_TYPE.htm) which you can get by calling [`SimConnect_GetInputEvent`](https://docs.flightsimulator.com/flighting/html/Programming_Tools/SimConnect/API_Reference/InputEvents/SimConnect_GetInputEvent.htm). According to `SimConnect.h` (look below) this is either `SIMCONNECT_INPUT_EVENT_TYPE_DOUBLE` (float64 / 8 bytes) or `SIMCONNECT_INPUT_EVENT_TYPE_STRING`. 

That's why I assume 8 bytes is correct for numeric values. As with the strings, in the notes for C# in the `SimConnect_GetInputEvent` docs, they refer to an `InputEventString` struct. After installing the SDK samples I found this by looking in the SimvarWatcher C# sample project:
```
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public class InputEventString
    {
        [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 256)]
        public string value;
    }
```
... which makes me think you are right that the strings has a fixed length (256?) 🤔 I haven't been able to test any string-based input events to see what works, buf if you do please let me know!

Just a headsup; the enums and structures shown in the online docs does not match what I find in SimConnect.h:

SIMCONNECT_INPUT_EVENT_DESCRIPTOR:
- [Online docs](https://docs.flightsimulator.com/flighting/html/Programming_Tools/SimConnect/API_Reference/Structures_And_Enumerations/SIMCONNECT_INPUT_EVENT_DESCRIPTOR.htm)
- SimConnect.h:
   ```c++
   SIMCONNECT_REFSTRUCT SIMCONNECT_INPUT_EVENT_DESCRIPTOR
   {
       SIMCONNECT_STRING(Name, 64);     // Input event name
       unsigned __int64	Hash;        // Hash
       SIMCONNECT_INPUT_EVENT_TYPE	eType;
   };
   ```
SIMCONNECT_INPUT_EVENT_TYPE:
- [Online docs](https://docs.flightsimulator.com/flighting/html/Programming_Tools/SimConnect/API_Reference/Structures_And_Enumerations/SIMCONNECT_INPUT_EVENT_TYPE.htm)
- SimConnect.h:
   ```c++
   SIMCONNECT_ENUM SIMCONNECT_INPUT_EVENT_TYPE : DWORD
   {
       SIMCONNECT_INPUT_EVENT_TYPE_DOUBLE,
       SIMCONNECT_INPUT_EVENT_TYPE_STRING
   };
   ```

**Reply by bstudtma** [reply]
The fix in v3.6.2 is working great. I'll keep an eye out for a string param. I didn't see one on the SR22T when I checked last night.. only noticed float64. I appreciate the quick fix!
